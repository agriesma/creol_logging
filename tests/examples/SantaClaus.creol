/*
 * SantaClaus.creol -- Santa claus synchronisation problem
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 * Original creol version by Ivar Alm
 *
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
interface SantaClausR
begin with Reindeer
  op backFromHoliday
end

interface SantaClausE
begin with Elf
  op haveProblem
end

interface SantaClaus inherits SantaClausR, SantaClausE
begin
end

interface Reindeer
begin
  with SantaClausR
    op harness
    op unharness
end

interface Elf
begin with SantaClausE
  // op enterOffice
  // op leaveOffice
  op showIn
  op showOut
end



class SantaClaus contracts SantaClaus
begin
  var ctReindeers: Int := 0
  var waitReindeers: List[Reindeer] := nil
  var harnessedReindeers: List[Reindeer] := nil
  var ctElves: Int := 0
  var waitElves: List[Elf] := nil
  var inofficeElves: List[Elf]  := nil

  op run == !loop()

  op loop == begin await ctReindeers = 5; deliverToys(;) []
              await  3 <= ctElves && ~ ctReindeers = 9; talkToElves(;) end ;
             !loop()

  op deliverToys ==
    var t1: Label[ ];
    var t2: Label[ ];
    var t3: Label[ ];
    var t4: Label[ ];
    var t5: Label[ ];
    ctReindeers := 0;
    t1!head(waitReindeers).harness(); 
    harnessedReindeers := harnessedReindeers |- head(waitReindeers); 
    waitReindeers := tail(waitReindeers);
    t2!head(waitReindeers).harness(); 
    harnessedReindeers := harnessedReindeers |- head(waitReindeers); 
    waitReindeers := tail(waitReindeers);
    t3!head(waitReindeers).harness(); 
    harnessedReindeers := harnessedReindeers |- head(waitReindeers); 
    waitReindeers := tail(waitReindeers);
    t4!head(waitReindeers).harness(); 
    harnessedReindeers := harnessedReindeers |- head(waitReindeers); 
    waitReindeers := tail(waitReindeers);
    t5!head(waitReindeers).harness(); 
    harnessedReindeers := harnessedReindeers |- head(waitReindeers); 
    waitReindeers := tail(waitReindeers);
    await t1? && t2? && t3? && t4? && t5?;
    // <Deliver Toys>;
    !head(harnessedReindeers).unharness();
    harnessedReindeers := tail(harnessedReindeers);
    // ....;
    !head(harnessedReindeers).unharness();
    harnessedReindeers := tail(harnessedReindeers)

  op talkToElves ==
        var t1: Label[ ];
        var t2: Label[ ];
        var t3: Label[ ];
	ctElves := ctElves - 3; 
	t1!head(waitElves).showIn();
	inofficeElves := inofficeElves |- head(waitElves);
	waitElves := tail(waitElves);
	t2!head(waitElves).showIn();
	inofficeElves := inofficeElves |- head(waitElves);
	waitElves := tail(waitElves);
	t3!head(waitElves).showIn();
	inofficeElves := inofficeElves |- head(waitElves);
	waitElves := tail(waitElves);
	await t1? && t2? && t3?;
	// <Talk to elves>;
	t1!head(inofficeElves).showOut();
	inofficeElves := tail(inofficeElves);
	t2!head(inofficeElves).showOut();
	inofficeElves := tail(inofficeElves);
	t3!head(inofficeElves).showOut();
	inofficeElves := tail(inofficeElves);
	await t1? && t2? && t3?

with Reindeer
  op backFromHoliday ==
	ctReindeers := ctReindeers + 1; 
	waitReindeers := waitReindeers |- caller

with Elf
  op haveProblem ==
	ctElves := ctElves + 1; 
	waitElves := waitElves |- caller
end


class Reindeer(sc: SantaClaus)
  contracts Reindeer
begin
  op run == !holiday()

  op holiday == /* <go on holiday>; */ !sc.backFromHoliday()

  op deliverToys == skip // <deliver Toys>

  with SantaClausR
    op harness == !deliverToys()
    op unharness == !holiday()
end



class Elf(sc:SantaClaus) contracts Elf
begin 
  op run == !work()
  op work == // <do work>;
		!sc.haveProblem()
  op talkToSanta == skip // <talk to Santa>

with SantaClausE
  op showIn == !talkToSanta()
  op showOut == !work()
end

class LeaderElf(sc:SantaClaus, elves: List[Elf])
  inherits Elf(sc)
  contracts Elf
begin
  op run == run:>Elf(;)
  op work ==
	// (<Lead elves> ||| <Make toys>);
	!sc.haveProblem()
end

class Christmas
begin
  op run == 
	var sc:SantaClaus;
	var r:Reindeer;
	var e:Elf;
	sc := new SantaClaus();
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	r := new Reindeer(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc);
	e := new Elf(sc)
end

//	#COMMENTS:
//
//	Local synchronous calls: m(...;...) is same as t!m(...;..); t?() //-
//	an abbrevation. 
//
//	!m(..)  || with no arguments  !m  starts the method call m but does
//	~ wait for any answer...
