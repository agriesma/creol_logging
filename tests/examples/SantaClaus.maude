load creol-interpreter .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op classes : -> Configuration [ctor] .
  eq classes =
    < "SantaClaus" : Class | Inh: noInh, Param: noVid, Att: "ctReindeers" |->
      null, "waitReindeers" |-> null, "harnessedReindeers" |-> null,
      "ctElves" |-> null, "waitElves" |-> null, "inofficeElves" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( ( "ctReindeers" @ "SantaClaus" ),
              ( "waitReindeers" @ "SantaClaus" ),
              ( "harnessedReindeers" @ "SantaClaus" ),
              ( "ctElves" @ "SantaClaus" ), ( "waitElves" @ "SantaClaus" ),
              ( "inofficeElves" @ "SantaClaus" ) ; int(0) :: list(emp) ::
              list(emp) :: int(0) :: list(emp) :: list(emp) ) ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:0" |-> null, "_" |->
              null, Code: call( "label:0" ; "this" ; "loop" ; emp ) ;
              free( "label:0" ) ; return ( emp ) >,
            < "loop" : Method | Param: noVid, Att: "label:2" |-> null,
              "label:1" |-> null, "label:0" |-> null, "_" |-> null,
              Code: (await "=" ( "ctReindeers" :: int(5) ) ;
                      call( "label:0" ; "this" ; "deliverToys" ; emp ) ;
                      get( "label:0" ; noVid ) ; free( "label:0" ) [] await
                      "&&" ( "<=" ( int(3) :: "ctElves" ) ::
                      "~" ( "=" ( "ctReindeers" :: int(9) ) ) ) ;
                      call( "label:1" ; "this" ; "talkToElves" ; emp ) ;
                      get( "label:1" ; noVid ) ; free( "label:1" ))
              ; call( "label:2" ; "this" ; "loop" ; emp ) ; free( "label:2" )
              ; return ( emp ) >,
            < "deliverToys" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "t1" |-> null, "t2" |-> null, "t3" |->
              null, "t4" |-> null, "t5" |-> null, "_" |-> null,
              Code: assign( "ctReindeers" ; int(0) ) ; call( "t1" ;
              "head" ( "waitReindeers" ) ; "harness" ; emp ) ;
              assign( "harnessedReindeers" ; "|-" ( "harnessedReindeers" ::
              "head" ( "waitReindeers" ) ) ) ;
              assign( "waitReindeers" ; "tail" ( "waitReindeers" ) ) ;
              call( "t2" ; "head" ( "waitReindeers" ) ; "harness" ; emp ) ;
              assign( "harnessedReindeers" ; "|-" ( "harnessedReindeers" ::
              "head" ( "waitReindeers" ) ) ) ;
              assign( "waitReindeers" ; "tail" ( "waitReindeers" ) ) ;
              call( "t3" ; "head" ( "waitReindeers" ) ; "harness" ; emp ) ;
              assign( "harnessedReindeers" ; "|-" ( "harnessedReindeers" ::
              "head" ( "waitReindeers" ) ) ) ;
              assign( "waitReindeers" ; "tail" ( "waitReindeers" ) ) ;
              call( "t4" ; "head" ( "waitReindeers" ) ; "harness" ; emp ) ;
              assign( "harnessedReindeers" ; "|-" ( "harnessedReindeers" ::
              "head" ( "waitReindeers" ) ) ) ;
              assign( "waitReindeers" ; "tail" ( "waitReindeers" ) ) ;
              call( "t5" ; "head" ( "waitReindeers" ) ; "harness" ; emp ) ;
              assign( "harnessedReindeers" ; "|-" ( "harnessedReindeers" ::
              "head" ( "waitReindeers" ) ) ) ;
              assign( "waitReindeers" ; "tail" ( "waitReindeers" ) ) ; await
              "&&" ( "&&" ( "&&" ( "&&" ( ?("t1") :: ?("t2") ) :: ?("t3") )
              :: ?("t4") ) :: ?("t5") ) ; free( "t5" ) ; free( "t4" ) ;
              free( "t3" ) ; free( "t2" ) ; free( "t1" ) ; call( "label:0" ;
              "head" ( "harnessedReindeers" ) ; "unharness" ; emp ) ;
              free( "label:0" ) ;
              assign( "harnessedReindeers" ; "tail" ( "harnessedReindeers" ) )
              ; call( "label:1" ; "head" ( "harnessedReindeers" ) ;
              "unharness" ; emp ) ; free( "label:1" ) ;
              assign( "harnessedReindeers" ; "tail" ( "harnessedReindeers" ) )
              ; return ( emp ) >,
            < "talkToElves" : Method | Param: noVid, Att: "t1" |-> null, "t2"
              |-> null, "t3" |-> null, "_" |-> null,
              Code: assign( "ctElves" ; "-" ( "ctElves" :: int(3) ) ) ;
              call( "t1" ; "head" ( "waitElves" ) ; "showIn" ; emp ) ;
              assign( "inofficeElves" ; "|-" ( "inofficeElves" ::
              "head" ( "waitElves" ) ) ) ;
              assign( "waitElves" ; "tail" ( "waitElves" ) ) ; call( "t2" ;
              "head" ( "waitElves" ) ; "showIn" ; emp ) ;
              assign( "inofficeElves" ; "|-" ( "inofficeElves" ::
              "head" ( "waitElves" ) ) ) ;
              assign( "waitElves" ; "tail" ( "waitElves" ) ) ; call( "t3" ;
              "head" ( "waitElves" ) ; "showIn" ; emp ) ;
              assign( "inofficeElves" ; "|-" ( "inofficeElves" ::
              "head" ( "waitElves" ) ) ) ;
              assign( "waitElves" ; "tail" ( "waitElves" ) ) ; await
              "&&" ( "&&" ( ?("t1") :: ?("t2") ) :: ?("t3") ) ; free( "t3" )
              ; free( "t2" ) ; free( "t1" ) ; call( "t1" ;
              "head" ( "inofficeElves" ) ; "showOut" ; emp ) ;
              assign( "inofficeElves" ; "tail" ( "inofficeElves" ) ) ;
              call( "t2" ; "head" ( "inofficeElves" ) ; "showOut" ; emp ) ;
              assign( "inofficeElves" ; "tail" ( "inofficeElves" ) ) ;
              call( "t3" ; "head" ( "inofficeElves" ) ; "showOut" ; emp ) ;
              assign( "inofficeElves" ; "tail" ( "inofficeElves" ) ) ; await
              "&&" ( "&&" ( ?("t1") :: ?("t2") ) :: ?("t3") ) ; free( "t3" )
              ; free( "t2" ) ; free( "t1" ) ; return ( emp ) >,
            < "backFromHoliday" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "ctReindeers" ; "+" ( "ctReindeers" :: int(1) ) )
              ; assign( "waitReindeers" ; "|-" ( "waitReindeers" ::
              "caller" ) ) ; return ( emp ) >,
            < "haveProblem" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "ctElves" ; "+" ( "ctElves" :: int(1) ) ) ;
              assign( "waitElves" ; "|-" ( "waitElves" :: "caller" ) ) ;
              return ( emp ) >,
      Ocnt: 0 >
    
    < "Reindeer" : Class | Inh: noInh, Param: "sc", Att: "sc" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:0" |-> null, "_" |->
              null, Code: call( "label:0" ; "this" ; "holiday" ; emp ) ;
              free( "label:0" ) ; return ( emp ) >,
            < "holiday" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null, Code: call( "label:0" ; "sc" ; "backFromHoliday" ;
              emp ) ; free( "label:0" ) ; return ( emp ) >,
            < "deliverToys" : Method | Param: noVid, Att: "_" |-> null,
              Code: skip ; return ( emp ) >,
            < "harness" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null,
              Code: call( "label:0" ; "this" ; "deliverToys" ; emp ) ;
              free( "label:0" ) ; return ( emp ) >,
            < "unharness" : Method | Param: noVid, Att: "label:0" |-> null,
              "_" |-> null,
              Code: call( "label:0" ; "this" ; "holiday" ; emp ) ;
              free( "label:0" ) ; return ( emp ) >,
      Ocnt: 0 >
    
    < "Elf" : Class | Inh: noInh, Param: "sc", Att: "sc" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:0" |-> null, "_" |->
              null, Code: call( "label:0" ; "this" ; "work" ; emp ) ;
              free( "label:0" ) ; return ( emp ) >,
            < "work" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null, Code: call( "label:0" ; "sc" ; "haveProblem" ; emp )
              ; free( "label:0" ) ; return ( emp ) >,
            < "talkToSanta" : Method | Param: noVid, Att: "_" |-> null,
              Code: skip ; return ( emp ) >,
            < "showIn" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null,
              Code: call( "label:0" ; "this" ; "talkToSanta" ; emp ) ;
              free( "label:0" ) ; return ( emp ) >,
            < "showOut" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null, Code: call( "label:0" ; "this" ; "work" ; emp ) ;
              free( "label:0" ) ; return ( emp ) >,
      Ocnt: 0 >
    
    < "LeaderElf" : Class | Inh: "Elf" < emp >, Param: "sc", "elves",
      Att: "sc" |-> null, "elves" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:0" |-> null, "_" |->
              null, Code: static( "label:0" ; "run" ; "Elf" ; "" ; emp ) ;
              get( "label:0" ; noVid ) ; free( "label:0" ) ; return ( emp ) >,
            < "work" : Method | Param: noVid, Att: "label:0" |-> null, "_"
              |-> null, Code: call( "label:0" ; "sc" ; "haveProblem" ; emp )
              ; free( "label:0" ) ; return ( emp ) >,
      Ocnt: 0 >
    
    < "Christmas" : Class | Inh: noInh, Param: noVid, Att: noSubst,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "sc" |-> null, "r" |->
              null, "e" |-> null, "_" |-> null,
              Code: new( "sc" ; "SantaClaus" ; emp ) ;
              new( "r" ; "Reindeer" ; "sc" ) ; new( "r" ; "Reindeer" ; "sc" )
              ; new( "r" ; "Reindeer" ; "sc" ) ;
              new( "r" ; "Reindeer" ; "sc" ) ; new( "r" ; "Reindeer" ; "sc" )
              ; new( "e" ; "Elf" ; "sc" ) ; new( "e" ; "Elf" ; "sc" ) ;
              new( "e" ; "Elf" ; "sc" ) ; new( "e" ; "Elf" ; "sc" ) ;
              new( "e" ; "Elf" ; "sc" ) ; new( "e" ; "Elf" ; "sc" ) ;
              new( "e" ; "Elf" ; "sc" ) ; return ( emp ) >,
      Ocnt: 0 > .
endm

