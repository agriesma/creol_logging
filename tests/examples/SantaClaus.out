==========================================
reduce in PROGRAM : classes .
result Configuration: 
< "Christmas" : Class | Inh: noInh, Param: noVid, Att: noSubst, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "run" : Method | Param: noVid, Att: "_" |-> null, "e" |-> null, "r"
    |-> null, "sc" |-> null, Code: new("sc" ; "SantaClaus" ; emp) ; new("r" ;
    "Reindeer" ; "sc") ; new("r" ; "Reindeer" ; "sc") ; new("r" ; "Reindeer" ;
    "sc") ; new("r" ; "Reindeer" ; "sc") ; new("r" ; "Reindeer" ; "sc") ; new(
    "e" ; "Elf" ; "sc") ; new("e" ; "Elf" ; "sc") ; new("e" ; "Elf" ; "sc") ;
    new("e" ; "Elf" ; "sc") ; new("e" ; "Elf" ; "sc") ; new("e" ; "Elf" ; "sc")
    ; new("e" ; "Elf" ; "sc") ; return(emp) >,
  Ocnt: 0 >
 
< "Elf" : Class | Inh: noInh, Param: "sc", Att: "sc" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "run" : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null,
    Code: call("label:0" ; "this" ; "work" ; emp) ; free("label:0") ; return(
    emp) >, < "showIn" : Method | Param: noVid, Att: "_" |-> null, "label:0"
    |-> null, Code: call("label:0" ; "this" ; "talkToSanta" ; emp) ; free(
    "label:0") ; return(emp) >, < "showOut" : Method | Param: noVid, Att: "_"
    |-> null, "label:0" |-> null, Code: call("label:0" ; "this" ; "work" ; emp)
    ; free("label:0") ; return(emp) >, < "talkToSanta" : Method | Param: noVid,
    Att: "_" |-> null, Code: skip ; return(emp) >, < "work" : Method | Param:
    noVid, Att: "_" |-> null, "label:0" |-> null, Code: call("label:0" ; "sc" ;
    "haveProblem" ; emp) ; free("label:0") ; return(emp) >,
  Ocnt: 0 >
 
< "LeaderElf" : Class | Inh: "Elf" < emp >, Param: "sc", "elves", Att: "elves"
    |-> null, "sc" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "run" : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null,
    Code: static("label:0" ; "run" ; "Elf" ; None ; emp) ; get("label:0" ;
    noVid) ; free("label:0") ; return(emp) >, < "work" : Method | Param: noVid,
    Att: "_" |-> null, "label:0" |-> null, Code: call("label:0" ; "sc" ;
    "haveProblem" ; emp) ; free("label:0") ; return(emp) >,
  Ocnt: 0 >
 
< "Reindeer" : Class | Inh: noInh, Param: "sc", Att: "sc" |-> null, Mtds:
  < "deliverToys" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
    return(emp) >, < "harness" : Method | Param: noVid, Att: "_" |-> null,
    "label:0" |-> null, Code: call("label:0" ; "this" ; "deliverToys" ; emp) ;
    free("label:0") ; return(emp) >, < "holiday" : Method | Param: noVid, Att:
    "_" |-> null, "label:0" |-> null, Code: call("label:0" ; "sc" ;
    "backFromHoliday" ; emp) ; free("label:0") ; return(emp) >, < "init" :
    Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp) >, <
    "run" : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null, Code:
    call("label:0" ; "this" ; "holiday" ; emp) ; free("label:0") ; return(emp)
    >, < "unharness" : Method | Param: noVid, Att: "_" |-> null, "label:0" |->
    null, Code: call("label:0" ; "this" ; "holiday" ; emp) ; free("label:0") ;
    return(emp) >,
  Ocnt: 0 >
 
< "SantaClaus" : Class | Inh: noInh, Param: noVid, Att: "ctElves" |-> null,
    "ctReindeers" |-> null, "harnessedReindeers" |-> null, "inofficeElves" |->
    null, "waitElves" |-> null, "waitReindeers" |-> null, Mtds:
  < "backFromHoliday" : Method | Param: noVid, Att: "_" |-> null, Code: assign(
    "ctReindeers" ; "+"("ctReindeers" :: int(1))) ; assign("waitReindeers" ;
    "|-"("waitReindeers" :: "caller")) ; return(emp) >, < "deliverToys" :
    Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null, "label:1" |->
    null, "t1" |-> null, "t2" |-> null, "t3" |-> null, "t4" |-> null, "t5" |->
    null, Code: assign("ctReindeers" ; int(0)) ; call("t1" ; "head"(
    "waitReindeers") ; "harness" ; emp) ; assign("harnessedReindeers" ; "|-"(
    "harnessedReindeers" :: "head"("waitReindeers"))) ; assign("waitReindeers"
    ; "tail"("waitReindeers")) ; call("t2" ; "head"("waitReindeers") ;
    "harness" ; emp) ; assign("harnessedReindeers" ; "|-"("harnessedReindeers"
    :: "head"("waitReindeers"))) ; assign("waitReindeers" ; "tail"(
    "waitReindeers")) ; call("t3" ; "head"("waitReindeers") ; "harness" ; emp)
    ; assign("harnessedReindeers" ; "|-"("harnessedReindeers" :: "head"(
    "waitReindeers"))) ; assign("waitReindeers" ; "tail"("waitReindeers")) ;
    call("t4" ; "head"("waitReindeers") ; "harness" ; emp) ; assign(
    "harnessedReindeers" ; "|-"("harnessedReindeers" :: "head"(
    "waitReindeers"))) ; assign("waitReindeers" ; "tail"("waitReindeers")) ;
    call("t5" ; "head"("waitReindeers") ; "harness" ; emp) ; assign(
    "harnessedReindeers" ; "|-"("harnessedReindeers" :: "head"(
    "waitReindeers"))) ; assign("waitReindeers" ; "tail"("waitReindeers")) ;
    await "&&"("&&"("&&"("&&"(?("t1") :: ?("t2")) :: ?("t3")) :: ?("t4")) :: ?(
    "t5")) ; free("t5") ; free("t4") ; free("t3") ; free("t2") ; free("t1") ;
    call("label:0" ; "head"("harnessedReindeers") ; "unharness" ; emp) ; free(
    "label:0") ; assign("harnessedReindeers" ; "tail"("harnessedReindeers")) ;
    call("label:1" ; "head"("harnessedReindeers") ; "unharness" ; emp) ; free(
    "label:1") ; assign("harnessedReindeers" ; "tail"("harnessedReindeers")) ;
    return(emp) >, < "haveProblem" : Method | Param: noVid, Att: "_" |-> null,
    Code: assign("ctElves" ; "+"("ctElves" :: int(1))) ; assign("waitElves" ;
    "|-"("waitElves" :: "caller")) ; return(emp) >, < "init" : Method | Param:
    noVid, Att: "_" |-> null, Code: assign(("ctReindeers" @ "SantaClaus"), (
    "waitReindeers" @ "SantaClaus"), ("harnessedReindeers" @ "SantaClaus"), (
    "ctElves" @ "SantaClaus"), ("waitElves" @ "SantaClaus"), ("inofficeElves" @
    "SantaClaus") ; int(0) :: list(emp) :: list(emp) :: int(0) :: list(emp) ::
    list(emp)) ; return(emp) >, < "loop" : Method | Param: noVid, Att: "_" |->
    null, "label:0" |-> null, "label:1" |-> null, "label:2" |-> null, Code: (
    await "&&"("<="(int(3) :: "ctElves") :: "~"("="("ctReindeers" :: int(9))))
    ; call("label:1" ; "this" ; "talkToElves" ; emp) ; get("label:1" ; noVid) ;
    free("label:1")[]await "="("ctReindeers" :: int(5)) ; call("label:0" ;
    "this" ; "deliverToys" ; emp) ; get("label:0" ; noVid) ; free("label:0")) ;
    call("label:2" ; "this" ; "loop" ; emp) ; free("label:2") ; return(emp) >,
    < "run" : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null,
    Code: call("label:0" ; "this" ; "loop" ; emp) ; free("label:0") ; return(
    emp) >, < "talkToElves" : Method | Param: noVid, Att: "_" |-> null, "t1"
    |-> null, "t2" |-> null, "t3" |-> null, Code: assign("ctElves" ; "-"(
    "ctElves" :: int(3))) ; call("t1" ; "head"("waitElves") ; "showIn" ; emp) ;
    assign("inofficeElves" ; "|-"("inofficeElves" :: "head"("waitElves"))) ;
    assign("waitElves" ; "tail"("waitElves")) ; call("t2" ; "head"("waitElves")
    ; "showIn" ; emp) ; assign("inofficeElves" ; "|-"("inofficeElves" ::
    "head"("waitElves"))) ; assign("waitElves" ; "tail"("waitElves")) ; call(
    "t3" ; "head"("waitElves") ; "showIn" ; emp) ; assign("inofficeElves" ;
    "|-"("inofficeElves" :: "head"("waitElves"))) ; assign("waitElves" ;
    "tail"("waitElves")) ; await "&&"("&&"(?("t1") :: ?("t2")) :: ?("t3")) ;
    free("t3") ; free("t2") ; free("t1") ; call("t1" ; "head"("inofficeElves")
    ; "showOut" ; emp) ; assign("inofficeElves" ; "tail"("inofficeElves")) ;
    call("t2" ; "head"("inofficeElves") ; "showOut" ; emp) ; assign(
    "inofficeElves" ; "tail"("inofficeElves")) ; call("t3" ; "head"(
    "inofficeElves") ; "showOut" ; emp) ; assign("inofficeElves" ; "tail"(
    "inofficeElves")) ; await "&&"("&&"(?("t1") :: ?("t2")) :: ?("t3")) ; free(
    "t3") ; free("t2") ; free("t1") ; return(emp) >,
  Ocnt: 0 >

Bye.
